examples/lowner_john.jl

```diff
@@ -0,0 +1,227 @@
+#                 C is PSD
+#
+# which is equivalent to a mixed conic quadratic and semidefinite
+# programming problem.
+#
+#
+# The outer ellipsoidal approximation to a polytope given
+# as the convex hull of a set of points
+#
+#     S = conv{ x1, x2, ... , xm }
+#
+# minimizes the volume of the enclosing ellipsoid,
+#
+#   { x | || P*(x-c) ||_2 <= 1 }
+#
+# The volume is proportional to det(P)^{-1/n}, so the problem can
+# be solved as
+#
+#   minimize         t
+#   subject to       t       >= det(P)^(-1/n)
+#               || P*xi + c ||_2 <= 1,  i=1,...,m
+#                 P is PSD.
+#
+# References:
+# [1] "Lectures on Modern Optimization", Ben-Tal and Nemirovski, 2000.
+#
+
+using JuMP
+import MathOptInterface
+const MOI = MathOptInterface
+
+# Models the convex set 
+#
+#   S = { (x, t) \in R^n x R | x >= 0, t <= (x1 * x2 * ... * xn)^(1/n) }
+#
+function geometric_mean(m, x, t)
+    n = length(x)
+    if n == 1
+        @constraint(m,t-x[1] <= 0)
+    else
+        t2 = @variable(m)
+        @constraint(m,1.0*[t2; x[n]; t] in MOI.PowerCone(1-1.0/n))
+        geometric_mean(m,x[1:n-1], t2)
+    end
+end
+
+# Purpose: Models the hypograph of the n-th power of the
+# determinant of a positive definite matrix. See [1,2] for more details.
+#
+#   The convex set (a hypograph)
+#
+#   C = { (X, t) \in S^n_+ x R |  t <= det(X)^{1/n} },
+#
+#   can be modeled as the intersection of a semidefinite cone
+#
+#   [ X, Z; Z^T Diag(Z) ] >= 0  
+#
+#   and a number of rotated quadratic cones and affine hyperplanes,
+#
+#   t <= (Z11*Z22*...*Znn)^{1/n}  (see geometric_mean).
+function det_root(m,X,t)
+    n = floor(Int,sqrt(length(X)))
+
+    @variable(m,Y[1:2*n,1:2*n],PSD)
+    @constraint(m,reshape(Y[n+1:2*n,1:n] - Y[n+1:2*n,n+1:2*n],n*n) in MOI.Zeros(n*n))
+    @constraint(m,reshape(X - Y[1:n,1:n], n*n) in MOI.Zeros(n*n))
+
+    geometric_mean(m,[Y[i+n,i+n] for i in 1:n],t)
+end
+
+#  The inner ellipsoidal approximation to a polytope 
+#
+#     S = { x \in R^n | Ax < b }.
+#
+#  maximizes the volume of the inscribed ellipsoid,
+#
+#     { x | x = C*u + d, || u ||_2 <= 1 }.
+#
+#  The volume is proportional to det(C)^(1/n), so the
+#  problem can be solved as 
+#
+#    maximize         t
+#    subject to       t       <= det(C)^(1/n)
+#                || C*ai ||_2 <= bi - ai^T * d,  i=1,...,m
+#                C is PSD
+#
+#  which is equivalent to a mixed conic quadratic and semidefinite
+#  programming problem.
+function lownerjohn_inner(solver :: MOI.AbstractOptimizer,
+                          A      :: Array{Float64,2},
+                          b      :: Vector{Float64})
+    m,n = size(A)
+
+    model = Model(mode=JuMP.Direct, backend=solver)
+
+    @variable(model, t >= 0)
+    @variable(model, C[1:n,1:n], PSD)
+    @variable(model, d[1:n])
+
+    # [ b-A*d ; A*C ] in QCone
+
+    for i in 1:m
+        @constraint(model, [ b[i] - (A[i,:]' * d) ; (A[i,:]'*C)'] in MOI.SecondOrderCone(n+1))
+    end
+
+    det_root(model,C,t)
```

Comment:
It should work with the LogDetConeTriangle once https://github.com/JuliaOpt/JuMP.jl/pull/1362 is merged

Reply:
That sounds great! It would simplify it a bit. 

Also, it hardcodes the Mosek solver and uses `mode=JuMP.Direct`, which I'm not sure is so pretty. Is there a way to let JuMP choose the solver based on the model type?

Reply:
No, automatically choosing the solver was removed during the update to Julia v0.6.
See https://github.com/JuliaOpt/MathProgBase.jl/pull/151
The new approach is to let the user choose the solver and then try to automatically bridge the constraints to match the types of constraints supported by the solvers instead of automatically selecting a solver that supports the types of the constraints created by the user.

Reply:
Probably a more solid approach. Is there any consensus on what examples should do?

Reply:
I don't think so. Currently the examples pick a solver.
What I usually do with Jupyter is creating different cells. Each cell sets `optimizer` to a different value and the user can run the cell he prefers in order to choose the solver.
Cc @joaquimg